clang/lib/CConv/CheckedRegions.cpp

```diff
@@ -28,18 +30,112 @@ using namespace clang;
 
 bool CheckedRegionAdder::VisitCompoundStmt(CompoundStmt *S) { 
   llvm::FoldingSetNodeID Id;
+  ast_type_traits::DynTypedNode DTN = ast_type_traits::DynTypedNode::create(*S);
+
   S->Profile(Id, *Context, true);
   switch (Map[Id]) {
-    case IS_UNCHECKED: return true;
-    case IS_CHECKED:   
-                        auto Loc = S->getBeginLoc();
-                        Writer.InsertTextBefore(Loc, "_Checked ");
-                        return false;
+    case IS_UNCHECKED:
+      if(isParentChecked(DTN) && !isFunctionBody(S)) {
+        auto Loc = S->getBeginLoc();
+        Writer.InsertTextBefore(Loc, "_Unchecked ");
+      }
+      break;
+    case IS_CHECKED:
+      if(!isParentChecked(DTN)) {
+        auto Loc = S->getBeginLoc();
+        Writer.InsertTextBefore(Loc, "_Checked ");
+      }
+      break;
+    default: llvm_unreachable("Bad flag in CheckedRegionAdder");
   }
 
-  llvm_unreachable("Bad flag in CheckedRegionAdder");
+  for(const auto C : S->children()) {
+    CheckedRegionAdder SV(Context, Writer, Map);
+    SV.TraverseStmt(C);
+  }
+
+  return false;
+}
+
+
+bool CheckedRegionAdder::VisitIfStmt(IfStmt *IS) {
+  CheckedRegionAdder Then_v(Context, Writer, Map);
+  CheckedRegionAdder Else_v(Context, Writer, Map);
+
+  Then_v.TraverseStmt(IS->getThen());
+  Else_v.TraverseStmt(IS->getElse());
+
+  return false;
 }
 
+
+bool CheckedRegionAdder::VisitCallExpr(CallExpr *C) {
+  auto FD = C->getDirectCallee();
+  FoldingSetNodeID ID;
+  ast_type_traits::DynTypedNode DTN = ast_type_traits::DynTypedNode::create(*C);
+
+  C->Profile(ID, *Context, true);
```

Comment:
What does `Profile` do?

Reply:
Is it somehow setting `ID`? Otherwise I'm not seeing `ID` getting a non-initial value.

Reply:
Extracts the NodeId, which is used as a unique identifier.

Reply:
Yes it sets ID. See: [Profile](https://clang.llvm.org/doxygen/classclang_1_1Stmt.html#a9638f0d4336d168d98741225263c4003)